### ISSUE TITLE
Created object will be released after Injector leave the scope?

### ISSUE THREAD

**Issue by raindust** [google/fruit#78]
```cpp
#include <iostream>

struct X {
	virtual void f() = 0;
};

struct Y : public X {};

struct Z : public Y {
	INJECT(Z()) = default;
	void f() override {
	}
};

fruit::Component<X> getComponent() {
	return fruit::createComponent()
			.bind<X, Y>()
			.bind<Y, Z>();
}

X *getX() {
	fruit::Injector<X> injector(getComponent);
	X *x = injector.get<X*>();
        Z *z = dynamic_cast<Z *>(x);
        return z;
}

int main() {
    X *x = getX();
    Z *z = dynamic_cast<Z *>(x); //will throw here
    
    return 0;
}
```
Above is a simple example copied from unit test, the created instance of Z perform well in getX() method, while throw after leaved the method.
So should i always hold the inject object? Or there is another option?

**Comment by poletti-marco** [comment]
Hi, yes, that's correct.
The Injector object owns all injected objects (the ones injected from that Injector).

The typical use case is to create an injector in main or in another close-to-top-level function/method and not destroy it until the injected objects are no longer needed (which could be the end of the process).

What's your use case?
E.g. are you developing a library or an application?
Would you like a different X each time or the same one?

Note that if you always want the same you can have a static variable in a method that holds the injector. This might make sense if it's a top-level function in a library, but if you do this elsewhere it might just make things more complicated.

If you haven't already, see the tutorial for some examples on how to use Fruit in a full system.

**Comment by raindust** [comment]
Hi @poletti-marco. Thanks for your answer.

I want to use it in a plugin based system, which creates component dynamically whenever it needs. 

I misunderstand the Injector class as a kind of function method. I seems a static injector variable would meet my needs :)

**Comment by poletti-marco** [comment]
Ok.
Do you know the entire set of possible plugins at start-up (and you just
don't want to load them until needed) or do you not have this information
at start-up?

If you do, you might want to install all components for all possible
plugins in a single injector created at start-up, and use Fruit's lazy
injection to only instantiate the plugin if/when needed.

If you don't it's a bit of a mess I'm afraid. You can have 1 static
injector per plugin, however note that they will be independent. So if both
plugins A and B inject an instance of a class X, there will be 1 per plugin
instead of only 1 as you might want/expect.
(or maybe this isolation is desirable/ok in your case?)

**Comment by raindust** [comment]
Hi @poletti-marco.
Yes, i actually know the entire set of possible plugins at start-up, and i originally planed to hold a set of "plugin objects", each of these objects has an injector independentlly. As you see, it can work but is ugly.

After you said "install all components for all possible plugins in a single injector created at start-up, and use Fruit's lazy injection to only instantiate the plugin if/when needed", i realized that i have missed some amazing features. Thank you for helping me find an elegant way to solve this.

So, i want to go deep into is aspect, can you give me some example links or point out which part of source should i start with (I read wiki samples before and did not find any clues)? Any suggestion would be greatly appreciated.

**Comment by poletti-marco** [comment]
https://github.com/google/fruit/wiki/quick-reference
Here you can find all Fruit features.

Specifically you should look at this:
https://github.com/google/fruit/wiki/quick-reference#lazy-injection

TL;DR in your non-plugin classes (say, X) that depend on plugins you should
always inject the plugin classes as Provider<SomePlugin>.
That way Fruit will be able to create an X without really instantiating the
plugin nor the stuff the plugin class depends on.
Then when you know you need the plugin (probably in a method of X) you can
call get on the Provider to actually get an instance (which will
instantiate the plugin if it hasn't been instantiated yet).

Elsewhere you can inject things normally, e.g. SomePlugin can inject a
HeavyClass directly (not via Provider) and HeavyClass won't be injected
until it's needed to construct SomePlugin (or something else).

If it's not clear, feel free to ask and I can explain further.

**Comment by raindust** [comment]
Hi @poletti-marco. That's so kind of you, thanks a lot. I will learn these first :)

**Comment by raindust** [comment]
Hi @poletti-marco, sorry to bother you again. I'm trying to create plugin with factories and can't figure out whether or not what i'm doing is a good practice.

Let's just suppose we are going to write a drawing system, and there are two basic interfaces about drawing element:

```cpp
class ILine {
public:
	virtual ~ILine() {}
};

class IArea {
public:
	virtual ~IArea() {}
};
```

and with one simple native definition respectively

```cpp
class Segment : public ILine {
};

class Plane : public IArea {
};
```

Note that these element will be created many times and may be shared (for example two faces of a cube may share the same line), so i use factory component to generate this.

```cpp
Component<std::function<LinePtr()>> getSegmentFactoryComponent() {
	return fruit::createComponent().registerFactory<LinePtr()>(
			[]() {
				return LinePtr(new Segment);
			});
}

Component<std::function<AreaPtr()>> getPlaneFactoryComponent() {
	return fruit::createComponent().registerFactory<AreaPtr()>(
			[]() {
				return AreaPtr(new Plane);
			});
}
```

With these two factories we can define a rectangle plugin component as follows.

```cpp
Component<std::function<AreaPtr()>, std::function<LinePtr()>> getRectanglePluginComponent() {
	return fruit::createComponent()
			.install(getPlaneFactoryComponent)
			.install(getSegmentFactoryComponent);
};
```

Then we define a plugin for creating circle and a sub class of ILine, which may be in a another plugin project.

```cpp
class Arc : public ILine {
};

Component<std::function<LinePtr()>> getArcFactoryComponent() {
	return fruit::createComponent().registerFactory<LinePtr()>(
			[]() {
				return LinePtr(new Arc);
			});
}

Component<std::function<AreaPtr()>, std::function<LinePtr()>> getCirclePluginComponent() {
	return fruit::createComponent()
			.install(getPlaneFactoryComponent)
			.install(getArcFactoryComponent);
};
```

And also a plugin for creating sphere and related IArea sub class.

```cpp
class Surface : public IArea {
};

Component<std::function<AreaPtr()>> getSurfaceFactoryComponent() {
	return fruit::createComponent().registerFactory<AreaPtr()>(
			[]() {
				return AreaPtr(new Surface);
			});
}

Component<std::function<AreaPtr()>, std::function<LinePtr()>> getSpherePluginComponent() {
	return fruit::createComponent()
			.install(getSurfaceFactoryComponent)
			.install(getArcFactoryComponent);
};
```

Finally, we defined a drawer class to create objects of ILine and IArea.  As you can see, this drawer class is a simple demostration with local variable to hold Injector object and simple conditional statement to create objects. In actual implementation, we can scan these plugin components and define a map of Injector objects to hold them (This is so far i can see).

```cpp
class Drawer {
public:
	void draw(const std::string &type) {
		typedef Injector<std::function<AreaPtr()>, std::function<LinePtr()>> Plugin;
		typedef std::shared_ptr<Plugin> PluginPtr;

		PluginPtr plugin;
		if (type == "Rectangle")
			plugin = PluginPtr(new Plugin(getRectanglePluginComponent));
		else if (type == "Circle")
			plugin = PluginPtr(new Plugin(getCirclePluginComponent));
		else if (type == "Sphere")
			plugin = PluginPtr(new Plugin(getSpherePluginComponent));

		std::function<LinePtr()> lineFactory = plugin->get<std::function<LinePtr()>>();
		//create some lines

		std::function<AreaPtr()> areaFactory = plugin->get<std::function<AreaPtr()>>();
		//create some areas
	}
};
```



OK, so much for this. Allow me to clear up the requirements:

- We need multiple instances of some classes (such as ILine, we will create multiple objects of it)
- The implements of these classes are in different project (so we can't list all the factory components)
- The plugin mainly defined a combination of sub classes (or a kind of scheme), and the user decide to choose one of these combinations for specific usage

My questions are:

- Should i abstract a "plugin interface" or these plugin components will be enough (If it is enough, is there  any way i can define the fruit Provider for lazy injection)?
- Are these factory components lightweight enough to define injectors whenever we want to?
- Is holding plugin Injector object within Drawer a good practice, or there is a better way?



Well , this is just my simple implement, i will greately preciate if you can show me a better solution :)


In the end, post the entire code for convenience: 

```cpp
#include <fruit/fruit.h>

using fruit::Component;
using fruit::Injector;

class ILine {
public:
	virtual ~ILine() {}
};

typedef std::shared_ptr<ILine> LinePtr;

class Segment : public ILine {
};

class Arc : public ILine {
};

class IArea {
public:
	virtual ~IArea() {}
};

typedef std::shared_ptr<IArea> AreaPtr;

class Plane : public IArea {
};

class Surface : public IArea {
};

Component<std::function<AreaPtr()>> getPlaneFactoryComponent() {
	return fruit::createComponent().registerFactory<AreaPtr()>(
			[]() {
				return AreaPtr(new Plane);
			});
}

Component<std::function<AreaPtr()>> getSurfaceFactoryComponent() {
	return fruit::createComponent().registerFactory<AreaPtr()>(
			[]() {
				return AreaPtr(new Surface);
			});
}

Component<std::function<LinePtr()>> getSegmentFactoryComponent() {
	return fruit::createComponent().registerFactory<LinePtr()>(
			[]() {
				return LinePtr(new Segment);
			});
}

Component<std::function<LinePtr()>> getArcFactoryComponent() {
	return fruit::createComponent().registerFactory<LinePtr()>(
			[]() {
				return LinePtr(new Arc);
			});
}

Component<std::function<AreaPtr()>, std::function<LinePtr()>> getRectanglePluginComponent() {
	return fruit::createComponent()
			.install(getPlaneFactoryComponent)
			.install(getSegmentFactoryComponent);
};

Component<std::function<AreaPtr()>, std::function<LinePtr()>> getCirclePluginComponent() {
	return fruit::createComponent()
			.install(getPlaneFactoryComponent)
			.install(getArcFactoryComponent);
};

Component<std::function<AreaPtr()>, std::function<LinePtr()>> getSpherePluginComponent() {
	return fruit::createComponent()
			.install(getSurfaceFactoryComponent)
			.install(getArcFactoryComponent);
};

class Drawer {
public:
	void draw(const std::string &type) {
		typedef Injector<std::function<AreaPtr()>, std::function<LinePtr()>> Plugin;
		typedef std::shared_ptr<Plugin> PluginPtr;

		PluginPtr plugin;
		if (type == "Rectangle")
			plugin = PluginPtr(new Plugin(getRectanglePluginComponent));
		else if (type == "Circle")
			plugin = PluginPtr(new Plugin(getCirclePluginComponent));
		else if (type == "Sphere")
			plugin = PluginPtr(new Plugin(getSpherePluginComponent));

		std::function<LinePtr()> lineFactory = plugin->get<std::function<LinePtr()>>();
		//create some lines

		std::function<AreaPtr()> areaFactory = plugin->get<std::function<AreaPtr()>>();
		//create some areas
	}
};
```

**Comment by poletti-marco** [comment]
> Should i abstract a "plugin interface" or these plugin components will be enough

It's not strictly necessary but IMO it would be much cleaner.

> (If it is enough, is there any way i can define the fruit Provider for lazy injection)?

Switching to an injected `Drawer`, you could do something like:

```
struct IPlaneFactory {
  virtual AreaPtr createArea() = 0;
};

struct PlaneFactory : public IPlaneFactory {
  INJECT(PlaneFactory()) = default;
  AreaPtr createArea() {
    return AreaPtr(new Plane());
  }
};

Component<IPlaneFactory> getPlaneFactoryComponent() {
	return fruit::createComponent()
	    .bind<IPlaneFactory, PlaneFactory>();
}

struct ISurfaceFactory {
  virtual AreaPtr createArea() = 0;
};

struct SurfaceFactory : public ISurfaceFactory {
  INJECT(SurfaceFactory()) = default;
  AreaPtr createArea() {
    return AreaPtr(new Surface());
  }
};

Component<ISurfaceFactory> getSurfaceFactoryComponent() {
	return fruit::createComponent()
	    .bind<ISurfaceFactory, SurfaceFactory>();
}

struct ISegmentFactory {
  virtual LinePtr createLine() = 0;
};

struct SegmentFactory : public ISegmentFactory {
  INJECT(SegmentFactory()) = default;
  LinePtr createLine() {
    return LinePtr(new Segment());
  }
};

Component<ISegmentFactory> getSegmentFactoryComponent() {
	return fruit::createComponent()
	    .bind<ISegmentFactory, SegmentFactory>();
}

struct IArcFactory {
  virtual LinePtr createLine() = 0;
};

struct ArcFactory : public IArcFactory {
  INJECT(ArcFactory()) = default;
  LinePtr createLine() {
    return LinePtr(new Arc());
  }
};

Component<IArcFactory> getArcFactoryComponent() {
	return fruit::createComponent()
	    .bind<IArcFactory, ArcFactory>();
}

struct IShape {
  virtual LinePtr createLine() = 0;
  virtual AreaPtr createArea() = 0;
};

struct RectangleTag {};

struct Rectangle : public IShape {
  IPlaneFactory* planeFactory;
  ISegmentFactory* segmentFactory;
  INJECT(Rectangle(IPlaneFactory* planeFactory, ISegmentFactory* segmentFactory))
      : planeFactory(planeFactory), segmentFactory(segmentFactory) {}
  
  LinePtr createLine() {
    return segmentFactory->createLine();
  }
  
  AreaPtr createArea() {
    return planeFactory->createArea();
  }
};

Component<fruit::Annotated<RectangleTag, IShape>> getRectanglePluginComponent() {
	return fruit::createComponent()
	    .bind<fruit::Annotated<RectangleTag, IShape>, Rectangle>()
        .install(getPlaneFactoryComponent)
        .install(getSegmentFactoryComponent);
};

struct CircleTag {};

struct Circle : public IShape {
  IPlaneFactory* planeFactory;
  IArcFactory* arcFactory;
  INJECT(Circle(IPlaneFactory* planeFactory, IArcFactory* arcFactory))
      : planeFactory(planeFactory), arcFactory(arcFactory) {}
  
  LinePtr createLine() {
    return arcFactory->createLine();
  }
  
  AreaPtr createArea() {
    return planeFactory->createArea();
  }
};

Component<fruit::Annotated<CircleTag, IShape>> getCirclePluginComponent() {
	return fruit::createComponent()
	    .bind<fruit::Annotated<CircleTag, IShape>, Circle>()
        .install(getPlaneFactoryComponent)
        .install(getArcFactoryComponent);
};

struct SphereTag {};

struct Sphere : public IShape {
  ISurfaceFactory* surfaceFactory;
  IArcFactory* arcFactory;
  INJECT(Sphere(ISurfaceFactory* surfaceFactory, IArcFactory* arcFactory))
      : surfaceFactory(surfaceFactory), arcFactory(arcFactory) {}
  
  LinePtr createLine() {
    return arcFactory->createLine();
  }
  
  AreaPtr createArea() {
    return surfaceFactory->createArea();
  }
};

Component<fruit::Annotated<SphereTag, IShape>> getSpherePluginComponent() {
	return fruit::createComponent()
	    .bind<fruit::Annotated<SphereTag, IShape>, Sphere>()
        .install(getSurfaceFactoryComponent)
        .install(getArcFactoryComponent);
};

struct IDrawer {
  void draw(const std::string &type);
};

struct Drawer : public IDrawer {
  Provider<IShape> rectangleProvider;
  Provider<IShape> circleProvider;
  Provider<IShape> sphereProvider;
  
  INJECT(Drawer(
      ANNOTATED(Rectangle, Provider<IShape>) rectangleProvider,
      ANNOTATED(Circle, Provider<IShape>) circleProvider,
      ANNOTATED(Sphere, Provider<IShape>) sphereProvider)) 
      : rectangleProvider(rectangleProvider), circleProvider(circleProvider), sphereProvider(sphereProvider) {}

  void draw(const std::string &type) {
      IShape* shape;
      if (type == "Rectangle")
        shape = rectangleProvider.get();
      else if (type == "Circle")
        shape = circleProvider.get();
      else if (type == "Sphere")
        shape = sphereProvider.get();
      else
        // This should never happen
        abort();

      LinePtr line = shape->createLine();
      // ...
      //create some lines

      AreaPtr area = shape->createArea();
      // ...
      //create some areas
  }
};

Component<IDrawer> getDrawerComponent() {
  return fruit::createComponent()
      .bind<IDrawer, Drawer>()
      .install(getRectanglePluginComponent)
      .install(getCirclePluginComponent)
      .install(getSpherePluginComponent);
}

int main() {
  Injector<IDrawer> injector(getDrawerComponent);
  IDrawer* drawer = injector.get<IDrawer*>();
  std::vector<std::string> shapes = {"Rectangle", "Rectangle", "Circle"};
  for (std::string shape : shapes) {
    drawer->draw(shape);
  }
}
```


> Are these factory components lightweight enough to define injectors whenever we want to?

Fruit tries to be efficient, but the injector construction is not free.
I would avoid creating multiple injectors unless it's really needed because you want 1 separate instance of every class for each injector. If repeated injector creation is required, Fruit has a `NormalizedComponent` class that would help save some (but not all) of the cost of repeated injector creation.
But here it doesn't look like multi-injector is needed, it would just complicate things and cause a performance hit.

> Is holding plugin Injector object within Drawer a good practice, or there is a better way?

As mentioned previously, the best practice is to create a single injector object for the entire process if it's enough (in production code) or 1 per test (in tests).

**Comment by raindust** [comment]
Hi @poletti-marco.  I was too focus on the combination of fruit components to realize that factory class can do these better, thanks a lot for your reply. I will go on to finish my plugin system :)
